sql_generators/glean_usage/glean_app_ping_views.py

```diff
@@ -256,3 +261,18 @@ def _generate_select_expression(
                         select_expr.append(f"SAFE_CAST(NULL AS {dtype}) AS {node_name}")
 
         return ", ".join(select_expr)
+
+    def _type_info(self, nodes):
+        """Determine the type information"""
+        type_expr = []
+        for node in nodes:
+            dtype = node["type"]
+            if dtype == "RECORD":
+                if node.get("mode", None) == "REPEATED":
+                    type_expr.append(f"ARRAY<STRUCT<{self._type_info(node['fields'])}>>")
+                else:
+                    type_expr.append(f"STRUCT<{self._type_info(node['fields'])}>")
+            else:
+                type_expr.append(dtype)
+
+        return ", ".join(type_expr)
```

Comment:
this will produce results like STRUCT<STRING, INT64>, which bigquery allows, but it makes the fields anonymous, which is undesirable. UNION ALL uses the field names from the first query for all subsequent queries, so if this happens in the first query it will anonymize field names from subsequent queries.

Instead fields should be named, and that combined with the above comment makes it so that this should probably accept a single node.

Also, mode needs to be checked regardless of whether the node is a record.

```suggestion
    def _type_info(self, node):
        """Determine the type information"""
        dtype = node["type"]
        if dtype == "RECORD":
            dtype = (
                "STRUCT<"
                + ", ".join(
                    f"{field['name']} {self._type_info(field)}"
                    for field in node["fields"]
                )
                + ">"
            )
        if node.get("mode", None) == "REPEATED":
            return f"ARRAY<{dtype}>"
        return dtype
```

Reply:
This incorporates/supersedes my suggestions, so I'll mark mine as resolved.